cmd/zq/zq.go

```diff
@@ -96,15 +97,57 @@ func New(f *flag.FlagSet) (charm.Command, error) {
 	return c, nil
 }
 
+func compileFilter(p ast.Proc) (filter.Filter, error) {
+	if fproc, ok := p.(*ast.FilterProc); ok {
+		return filter.Compile(fproc.Filter)
+	}
+	return nil, nil
+}
+
+func liftFilter(p ast.Proc) (ast.Proc, filter.Filter, error) {
+	f, err := compileFilter(p)
+	if err != nil {
+		return nil, nil, err
+	}
+	if f != nil {
+		pass := &ast.PassProc{
+			Node: ast.Node{"PassProc"},
+		}
+		return pass, f, nil
+	}
+	seq, ok := p.(*ast.SequentialProc)
+	if ok && len(seq.Procs) > 0 {
+		f, err := compileFilter(seq.Procs[0])
+		if err != nil {
+			return nil, nil, err
+		}
+		if f != nil {
+			rest := &ast.SequentialProc{
+				Node:  ast.Node{"SequentialProc"},
+				Procs: seq.Procs[1:],
+			}
+			return rest, f, nil
+		}
+	}
+	return p, nil, nil
+}
+
 func (c *Command) compile(p ast.Proc, reader zson.Reader) (*proc.MuxOutput, error) {
 	ctx := &proc.Context{
 		Context:  context.Background(),
 		Resolver: resolver.NewTable(),
 		Logger:   zap.NewNop(),
 		Warnings: make(chan string, 5),
 	}
-	scr := scanner.NewScanner(reader)
-	leaves, err := proc.CompileProc(nil, p, ctx, scr)
+	// Try to move the filter into the scanner so we can throw
+	// out unmatched records without copying their contents.
```

Comment:
Clarify that "without copying their contents" is specific to raw?

Reply:
+1